pkg/command/context_tkg_helper.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package command
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	tkgauth "github.com/vmware-tanzu/tanzu-cli/pkg/auth/tkg"
+	wcpauth "github.com/vmware-tanzu/tanzu-cli/pkg/auth/wcp"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/pluginsupplier"
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/log"
+)
+
+const (
+	pinnipedAuthPluginName = "pinniped-auth"
+)
+
+type TKGKubeconfigFetcherOptions struct {
+	Endpoint              string
+	InsecureSkipTLSVerify bool
+	EndpointCACertPath    string
+	CmdExecutor           func(command *cobra.Command, args []string) (stdOut []byte, stderr []byte, err error)
+}
+
+type TKGKubeconfigFetcher interface {
+	GetPinnipedKubeconfig() (mergeFilePath, currentContext string, err error)
+}
+
+func NewTKGKubeconfigFetcher(endpoint, endpointCACertPath string, insecureSkipTLSVerify bool) TKGKubeconfigFetcher {
+	return &TKGKubeconfigFetcherOptions{
+		Endpoint:              endpoint,
+		EndpointCACertPath:    endpointCACertPath,
+		InsecureSkipTLSVerify: insecureSkipTLSVerify,
+		CmdExecutor:           RunCommandAndGetStdOutAndErr,
+	}
+}
+
+func (tkfo *TKGKubeconfigFetcherOptions) GetPinnipedKubeconfig() (string, string, error) {
+	exists, err := tkfo.isPinnipedPluginInstalled()
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to determine the 'pinniped-auth' plugin is installed")
+	}
+	if !exists {
+		return "", "", errors.New("'pinniped-auth' plugin is not installed. This is plugin is required to login to TKG/vSphere with Kubernetes(TKGs), please install the plugin and retry")
+	}
+
+	// get kubeconfig using the 'pinniped-auth' plugin. If the pinniped-auth plugin doesn't support `kubeconfig get` command,
+	// use the existing logic to fetch the kubeconfig for prior versions of TKG/TKGs
+	kubeconfigBytes, err := tkfo.getKubeconfigUsingPinnipedAuthPlugin()
+	if err != nil {
+		if strings.Contains(err.Error(), `unknown command "kubeconfig" for "pinniped-auth"`) {
```

Comment:
Considering we have a cobra command object, we might be able to use `cmd.Commands()` to get all subcommands and see if `kubeconfig` command exists or not.
cc @marckhouzam if there is a better way to do this with cobra library.

Reply:
Yes, I am looking into it.

Reply:
The root command is showing the `pinniped-auth` in rootcmd.Command(), but if i recursively scan the `pinniped-auth` commands( `Commands()` on pinniped-auth command) the list is showing empty instead of showing `kubeconfig` and `login`. cc: @marckhouzam

Reply:
This might be because the `pinniped-auth` plugin is a separate binary that tanzu-cli invokes and possibly the child commands are not available in a way that we are invoking plugin binaries separately.

Reply:
Yes, that's my thinking too.

Reply:
You guys are right.  In the CLI code, we only have a command for the top plugin level (`pinniped-auth`) but we don't know anything about its sub-commands.

There are different ways to figure out what you need to know, but they all require running the `pinniped-auth` binary.
In the end, doing it the way that is proposed here is just as good as any other approach, I think.